src/TorchSharp/JIT/ScriptModule.cs

```diff
@@ -163,9 +163,63 @@ public override void eval()
                 [DllImport("LibTorchSharp")]
                 static extern void THSJIT_Module_to_device(HType module, long deviceType, long deviceIndex);
 
+                [DllImport("LibTorchSharp")]
+                static extern void THSJIT_Module_to_device_dtype(HType module, sbyte dtype, long deviceType, long deviceIndex);
+
                 [DllImport("LibTorchSharp")]
                 static extern void THSJIT_Module_to_dtype(HType module, sbyte dtype);
 
+                public override nn.Module to(Device device, ScalarType dtype)
+                {
+                    if (device.type != DeviceType.CUDA) { device = new Device(device.type, -1); };
+
+                    if (device.type == DeviceType.CUDA && !torch.cuda.is_available()) throw new InvalidOperationException("CUDA is not available.");
+
+                    if (device.type != _deviceType || device.index != _deviceIndex) {
+
+                        InitializeDeviceType(device.type);
+                        THSJIT_Module_to_device_dtype(handle, (sbyte)dtype, (int)device.type, device.index);
+                        CheckForErrors();
+
+                        foreach (var (_, sm) in named_children()) sm.to(device.type, device.index);
+
+                        foreach (var field in GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)) {
+
+                            var fieldName = field.Name;
+                            var value = field.GetValue(this);
+
+                            switch (value) {
+                            // This test must come before the Tensor test
+                            case Modules.Parameter param when dtype == param.dtype && device.type == param.device_type && device.index == param.device_index:
+                                continue;
+
+                            case Modules.Parameter param: {
+                                    var t = param.to(dtype, device);
+                                    t.retain_grad();
```

Comment:
Why is this needed? Should it only be called when the original parameter `requires_grad`?

Reply:
Parameters should always require gradients.